src/compiler/sys.ts

```diff
@@ -298,6 +298,55 @@ namespace ts {
                 };
             }
 
+            function createWatchedFileSet() {
+                const watchedDirectories = createFileMap<FileWatcher>();
+                const watchedFiles = createFileMap<(fileName: string, removed?: boolean) => void>();
+                const currentDirectory = process.cwd();
+
+                return { addFile, removeFile };
+
+                function addFile(fileName: string, callback: (fileName: string, removed?: boolean) => void): WatchedFile {
+                    const path = toPath(fileName, currentDirectory, getCanonicalPath);
+                    const parentDirPath = toPath(ts.getDirectoryPath(fileName), currentDirectory, getCanonicalPath);
+
+                    if (!watchedDirectories.contains(parentDirPath)) {

```

Comment:
So you add a directory watcher for every directory that contains files being watched? Seems like if we are already watching the directory containing `tsconfig.json`, and you can watch a directory recursively, why would you ever need to watch a subdirectory of the project root? (i.e. if you watch a parent dir recursively, you should never need to also watch a child dir).